### PR TITLE
[release-4.10] OCPBUGS-2013: Validate node subnet annotations against cluster networks

### PR DESCRIPTION
**Pull request by pperiyasamy**
This is manual cherry-pick of 4.11 commits a11e45dfc49537e506c661410b37b5285b0a893f 2d28afb74a330515f121e776f94de60b1000c1fe 711237766b25e0728006eacb8dc1478b1ce359bb 484c5a002c80c5b3dca1b67194dc5271ba46acce 3eb7a5e7bc8e2dc7b8fef4762dd96b1cd20c6fd4.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

### CODE REVIEW

**Comment by dcbw** [comment]
/hold

Windows team reported regression in the 4.11 backport which I think this PR will also be subject to. @JacobTanenbaum is figuring out the fix.

**Comment by dcbw** [comment]
I think we can unhold this now that it's fixed...

/hold cancel

**Comment by dcbw** [comment]
/approve

@JacobTanenbaum can you LGTM if it looks OK to you now?

File: go-controller/pkg/ovn/ovn.go [hunk 1]
```diff
@@ -1362,8 +1362,7 @@ func (oc *Controller) WatchNodes() {
 			klog.V(5).Infof("Delete event for Node %q. Removing the node from "+
 				"various caches", node.Name)
 
-			nodeSubnets, _ := util.ParseNodeHostSubnetAnnotation(node)
-			oc.deleteNode(node.Name, nodeSubnets)
+			oc.deleteNode(node.Name)
```
Comment (jcaamano) [inline go-controller/pkg/ovn/ovn.go:1365]: Should we Klog something here on error?

File: go-controller/pkg/ovn/master.go [hunk 2]
```diff
@@ -1236,15 +1097,7 @@ func (oc *Controller) deleteNodeLogicalNetwork(nodeName string) error {
...
-	for _, hostSubnet := range hostSubnets {
-		if err := oc.deleteNodeHostSubnet(nodeName, hostSubnet); err != nil {
-			klog.Errorf("Error deleting node %s HostSubnet %v: %v", nodeName, hostSubnet, err)
-		} else {
-			util.UpdateUsedHostSubnetsCount(hostSubnet, &oc.v4HostSubnetsUsed, &oc.v6HostSubnetsUsed, false)
-		}
-	}
-	// update metrics
-	metrics.RecordSubnetUsage(oc.v4HostSubnetsUsed, oc.v6HostSubnetsUsed)
+	oc.masterSubnetAllocator.ReleaseNodeSubnets(nodeName, hostSubnets...)
```
[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/ovn/master.go]: OK, I see that a later commit changes this to not do anything on error. All fine (although having a klog warning would have been reasonable as well).

**Review by jcaamano — COMMENTED** [review]
Two things in addition to the comments:

* For commit `[ovn: ensure existing node subnet annotations are valid when a node is added](https://github.com/openshift/ovn-kubernetes/pull/1332/commits/ed408a58c6c76feff6bf7e851170ff916c8e773e)` I am missing some changes in `master_test.go`. Is that intentional?
* Commit https://github.com/openshift/ovn-kubernetes/pull/1332/commits/79141ba7f4ac2ecdf60c931a0f68edfea8684845 has also nothing to do with the cherry-picked commit, its completely different and unrelated. So I am missing something here as well.

**Comment by pperiyasamy** [comment]
> Two things in addition to the comments:
> 
> * For commit `[ovn: ensure existing node subnet annotations are valid when a node is added](https://github.com/openshift/ovn-kubernetes/pull/1332/commits/ed408a58c6c76feff6bf7e851170ff916c8e773e)` I am missing some changes in `master_test.go`. Is that intentional?
> * Commit [79141ba](https://github.com/openshift/ovn-kubernetes/commit/79141ba7f4ac2ecdf60c931a0f68edfea8684845) has also nothing to do with the cherry-picked commit, its completely different and unrelated. So I am missing something here as well.

Yes @jcaamano The commit for [79141ba](https://github.com/openshift/ovn-kubernetes/commit/79141ba7f4ac2ecdf60c931a0f68edfea8684845) doesn't have changes to master.go because on 4.10 it doesn't have `addUpdateNodeEvent` implementation, hence ignored `recordNodeErrorEvent` changes and also used it as placeholder to get required fixes in master_test.go (as per refactoring done in 4.11).  klog and other stuffs just came up while cherry-picking actual 4.11 commits.

File: go-controller/pkg/ovn/master_test.go [hunk 3]
```diff
@@ -1217,289 +1216,68 @@ var _ = ginkgo.Describe("Gateway Init Operations", func() {
...
+		ExternalIP: externalIP,
+		Type:       nbdb.NATTypeSNAT,
+		Options: map[string]string{
+			"stateless": "false",
+		},
 	}
 }
 
-func TestController_syncNodesRetriable(t *testing.T) {
+func TestController_syncNodes(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply go-controller/pkg/ovn/master_test.go]: These changes came from 4.11 at cherry-pick, removed it now.

File: go-controller/pkg/ovn/master.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply go-controller/pkg/ovn/master.go]: yes, done.

**Comment by pperiyasamy** [comment]
> > Two things in addition to the comments:
> > 
> > * For commit `[ovn: ensure existing node subnet annotations are valid when a node is added](https://github.com/openshift/ovn-kubernetes/pull/1332/commits/ed408a58c6c76feff6bf7e851170ff916c8e773e)` I am missing some changes in `master_test.go`. Is that intentional?
> > * Commit [79141ba](https://github.com/openshift/ovn-kubernetes/commit/79141ba7f4ac2ecdf60c931a0f68edfea8684845) has also nothing to do with the cherry-picked commit, its completely different and unrelated. So I am missing something here as well.
> 
> Yes @jcaamano The commit for [79141ba](https://github.com/openshift/ovn-kubernetes/commit/79141ba7f4ac2ecdf60c931a0f68edfea8684845) doesn't have changes to master.go because on 4.10 it doesn't have `addUpdateNodeEvent` implementation, hence ignored `recordNodeErrorEvent` changes and also used it as placeholder to get required fixes in master_test.go (as per refactoring done in 4.11). klog and other stuffs just came up while cherry-picking actual 4.11 commits.



> > Two things in addition to the comments:
> > 
> > * For commit `[ovn: ensure existing node subnet annotations are valid when a node is added](https://github.com/openshift/ovn-kubernetes/pull/1332/commits/ed408a58c6c76feff6bf7e851170ff916c8e773e)` I am missing some changes in `master_test.go`. Is that intentional?
> > * Commit [79141ba](https://github.com/openshift/ovn-kubernetes/commit/79141ba7f4ac2ecdf60c931a0f68edfea8684845) has also nothing to do with the cherry-picked commit, its completely different and unrelated. So I am missing something here as well.
> 
> Yes @jcaamano The commit for [79141ba](https://github.com/openshift/ovn-kubernetes/commit/79141ba7f4ac2ecdf60c931a0f68edfea8684845) doesn't have changes to master.go because on 4.10 it doesn't have `addUpdateNodeEvent` implementation, hence ignored `recordNodeErrorEvent` changes and also used it as placeholder to get required fixes in master_test.go (as per refactoring done in 4.11). klog and other stuffs just came up while cherry-picking actual 4.11 commits.

@jcaamano For 1, rebased the commit https://github.com/openshift/ovn-kubernetes/pull/1332/commits/19159d9f9b63652aa0080cb58bb1f2432326ee74 to have those missing test changes. For 2, Dropped out cherry-pick of https://github.com/openshift/ovn-kubernetes/commit/3eb7a5e7bc8e2dc7b8fef4762dd96b1cd20c6fd4 commit to this PR as it's not needed for 4.10. Hope it makes sense now.

**Pull request merged**